zilliqa/src/consensus.rs

```diff
@@ -568,6 +580,11 @@ impl Consensus {
                 if let Some(e) = e.downcast_ref::<MissingBlockError>() {
                     warn!("missing finalized block1");
                     info!(?e, "missing block when checking block proposal");
+
+                    let random_peer = self.get_random_other_peer();
+                    self.block_store
+                        .request_blocks(random_peer, block.header.number.saturating_sub(1))?;
```

Comment:
Should we be requesting the block from the `BlockRef` in the `MissingBlockError`? Why are we sure that `block.header.number.saturating_sub(1)` is the missing one?

Reply:
Most of the time this is correct, with edge cases being you are just joined/syncing. `request_blocks` also performs a check that the number is sane, and lowers it in the case it's above HEAD, so you should never request a block you clearly cannot attach to your chain.

Reply:
But `fn check_block` sometimes returns a `MissingBlockError` when the missing block isn't the `block`s parent. So doesn't this mean we'll end up unnecessarily requesting a block we already have and not requesting the block we're actually missing.

If we instead take the `BlockRef` out of `e`, then we can know exactly what was missing and request that?

Reply:
I didn't do it that way because if you allow someone to send you a block and you generate a network request on the specific missing information you could get into situations where you have two blocks that have missing refs to each other on the network and everything explodes into network requests. Or everyone requesting genesis's parent ref all the time or something.

Also, in the normal case where you join the network and start to sync, you will recieve a proposal from HEAD which may be very far in the future. You request the missing ref for HEAD-1 which makes you request the missing ref for HEAD-2... and so on, zipping down. This is super inefficient.

To make things as simplistic/minimalist as possible, I arranged block sync like follows:
- Only make a request if you have cause to believe you are missing information
- Only request from a single random peer
- This request ONLY refers to block number
  - block number should always reference a valid block (unlike view), unless it's above HEAD
  - the limited blocks number that are relevant to you to request are `> finalized block number`, and `<= HEAD + 1`
- The peer has the discresion to respond with any amount of blocks upward from this height
- If all the blocks you recieve from a request are valid and new, submit another request (zipping upwards)

Happy to discuss with you and @theo-zil whether you agree with these axioms or not - but it seems beyond the scope of this PR so I would suggest it's addressed in a follow up one

Reply:
I agree with you @n-hutton - specifically, the check in `block_store` to never request a block that's detached from your local chain ensures that the chain of requests for HEAD-1, HEAD-2... etc never happens.

But this raises another question: when would you ever want to request blocks at anything other than your current local highest_block_number? As I understand it, we're currently operating on the assumption that nodes will never need to request blocks earlier than its head. (For example, because every node holds a full database from genesis.) But at the same time, for aforementioned reasons you never want to request blocks above your head+1, either. So could we make `block_store.request_blocks()` take zero parameters, and have it effectively always just request a new batch of blocks starting from your head+1?

Happy to address this in a follow-up PR if you prefer. Or even migrate this discussion to slack if you want to merge this first.

Reply:
Hi Theo, it is not correct that nodes will not request blocks earlier than head.

In the happy case, there will be two blocks at height N+1 and N+2, which are not finalized, while N is finalized. If you recieve a block proposal which has a missing parent at height N+2 (HEAD), you should request from other nodes the block N+1. In the non-happy case, the number of non-finalized blocks can be > 2 and so I believe you could be in a very extreme case requesting far before HEAD.

There is also another case, which is when you recieve a valid block proposal that you personally reject because the view is less than your current view, but it later gets accepted by the rest of the network. This happens because it is impossible to avoid view desync in certain scenarios. This is tested in the forking test.

Reply:
I guess, now that I think about it, if we wanted to, it could be simplified in the way that you suggest by just requesting `finalized + 1`.